components/eamxx/src/diagnostics/extraaci.cpp

```diff
@@ -0,0 +1,255 @@
+
+  view_2d qt_v("qt_v", m_ncols, m_nlevs);
+  view_2d tl_v("tl_v", m_ncols, m_nlevs);
+  calc_tl_qt(tm_v, pm_v, qv_v, qc_v, tl_v, qt_v);
+
+  const auto &curr_ts =
+      get_field_in("qc").get_header().get_tracking().get_time_stamp();
+
+  std::int64_t dt = curr_ts - m_start_t;
+
+  const int num_levs  = m_nlevs;
+  const int pblinvalg = m_pblinvalg;
+  const auto policy   = ESU::get_default_team_policy(m_ncols, m_nlevs);
+
+  constexpr int wsms = 1;
+  using WSMgr        = ekat::WorkspaceManager<Real, DefaultDevice>;
+  WSMgr wsm(num_levs, wsms, policy);
+
+  Kokkos::parallel_for(
+      "Compute " + name(), policy, KOKKOS_LAMBDA(const MT &team) {
+        const int icol     = team.league_rank();
+        const auto qc_icol = ekat::subview(qc_v, icol);
+        const auto qv_icol = ekat::subview(qv_v, icol);
+        const auto tm_icol = ekat::subview(tm_v, icol);
+        const auto pm_icol = ekat::subview(pm_v, icol);
+        const auto pd_icol = ekat::subview(pd_v, icol);
+
+        const auto qt_icol = ekat::subview(qt_v, icol);
+        const auto tl_icol = ekat::subview(tl_v, icol);
+
+        auto prev_qtot_icol = ekat::subview(prev_qtot_v, icol);
+        auto prev_tliq_icol = ekat::subview(prev_tliq_v, icol);
+
+        auto ws = wsm.get_workspace(team);
+        ekat::Unmanaged<WSMgr::view_1d<Real>> tm_grad;
+        ws.take_many_contiguous_unsafe<wsms>({"tm_grad"}, {&tm_grad});
+
+        // We first want to find the PBL inversion. There are three methods to
+        // do so. All our methods here rely on the *gradient* of state fields
+        // (for now). First, we can simply find the first level from the surface
+        // that has a a temperature "inversion" (temperature goes positive
+        // instead of negative). Second, we can find the level which has the
+        // biggest positive jump in theta_l. Third, we can find the level which
+        // has the biggest negative jump in qt.
+
+        // We apply the first method only for now for now.
+        // We aim to find the minimum temperature gradient,
+        // indicating a positive temperature jump, but pressure
+        // is decreasing, so a negative gradient.
+        Real min_tm_grad    = 0.0;
+        int min_tm_grad_lev = 0;
+        if(pblinvalg == 1) {
+          // Find tm_grad = d(T_mid)/d(p_mid)
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 1, num_levs),
+                               [&](int k) {
+                                 tm_grad(k) = (tm_icol(k - 1) - tm_icol(k)) /
+                                              (pm_icol(k - 1) - pm_icol(k));
+                               });
+          team.team_barrier();
+
+          // Find minimum gradient, because d(p_mid) < 0 in definition above
+          // Starting from the surface, and ensuring p_mid > 70000.0 Pa,
+          // to avoid resolving to some odd place higher up in the atmosphere.
+          for(int k = num_levs; k > 1; k--) {
```

Comment:
This part of the code should be contained in a `Kokkos::single`:
```c++
Kokkos::singLe(Kokkos::PerTeam(team),[&]{
   for (...) {
     ...
   }
});
```

Technically, the way you wrote it, it works, since `min_tm_grad_lev` is an automatic var, which is thread-local, and all threads in the team will compute the same thing. Still...

Reply:
You could even think of using a `parallel_reduce`, if you feel like it, and avoid the single.